### PR TITLE
fix the bug with delegating optional concept functions

### PR DESCRIPTION
**Pull request by anstaf**
`delegate` helper was redesigned. Now it implements only those concept functions that are defined in the source SID. In the previous design all concept functions were defined (including those that are fall back to default for the source SID). As a result it was a lot of hard to debug compilation errors.

### CODE REVIEW

File: tests/unit_tests/sid/test_sid_rename_dimension.cpp [hunk 1]
```diff
@@ -52,5 +53,17 @@ namespace gridtools {
...
+        TEST(rename_dimensions, rename_twice_and_make_composite) {
+            double data[3][5][7];
+            auto src = sid::synthetic()
+                           .set<property::origin>(sid::make_simple_ptr_holder(&data[0][0][0]))
+                           .set<property::strides>(tu::make<hymap::keys<a, b, c>::values>(5_c * 7_c, 7_c, 1_c))
+                           .set<property::upper_bounds>(tu::make<hymap::keys<a, b>::values>(3, 5));
+            auto testee = sid::rename_dimension<a, c>(sid::rename_dimension<b, d>(src));
+            static_assert(sid::is_sid<decltype(testee)>(), "");
+            auto composite = tu::make<gridtools::sid::composite::keys<void>::values>(testee);
+            static_assert(sid::is_sid<decltype(composite)>(), "");
```
Comment (havogt) [inline tests/unit_tests/sid/test_sid_rename_dimension.cpp:66]:
```suggestion
            static_assert(sid::is_sid<decltype(composite)>(), "");
            auto ptrs = sid::get_origin(composite);
```

File: include/gridtools/sid/delegate.hpp [hunk 2]
```diff
@@ -25,36 +26,39 @@ namespace gridtools {
          * @tparam Sid a object that models `SID` concept.
```
Comment (havogt) [inline include/gridtools/sid/delegate.hpp:26]: Can you emphasize somewhere in the comments, e.g. here, that users have to be careful with value vs reference semantics in the delegates.

File: include/gridtools/sid/delegate.hpp [hunk 3]
```diff
@@ -25,36 +26,39 @@ namespace gridtools {
...
-
-            template <class Arg>
-            explicit delegate(Arg &&arg) noexcept : m_impl(std::forward<Arg>(arg)) {}
+            template <bool IsRef = std::is_reference<Sid>::value, std::enable_if_t<IsRef, int> = 0>
+            delegate(Sid impl) : m_impl(impl) {}
         };
 
         template <class Sid>
-        ptr_diff_type<Sid> sid_get_ptr_diff(delegate<Sid> const &);
+        decltype(sid_get_origin(std::declval<Sid &>())) sid_get_origin(delegate<Sid> &obj) {
```
Comment (havogt) [inline include/gridtools/sid/delegate.hpp:41]: It's easy to overlook that this works as a SFINAE-enabler. Maybe put a comment.

**Comment by havogt** [comment]
Also, please add in the commit message for merge what the problem was.

**Comment by anstaf** [comment]
launch jenkins

File: include/gridtools/sid/delegate.hpp (hunk 3 above)
Comment (anstaf) [reply include/gridtools/sid/delegate.hpp:41]: done

File: include/gridtools/sid/delegate.hpp (hunk 2 above)
Comment (anstaf) [reply include/gridtools/sid/delegate.hpp:26]: done

**Comment by anstaf** [comment]
launch jenkins

**Comment by anstaf** [comment]
launch jenkins

**Pull request merged**